### PR TITLE
Make setting textarea descriptions consistent with text fields (#51436)

### LINKED ISSUE #51436: [Enhancement]: Make descriptions in textarea fields consistent with text fields in WooCommerce settings
### Describe the solution you'd like

Where textarea fields are registered in WooCommerce settings the description (if not a tooltip) it displayed above the field which is inconsistent with text fields, which display the description below.

In addition, the text color of the description is a different shade compared to the description of most other fields.

The color should be consistent, there might be reasoning of why the description of textareas is placed above the field, but it potentially just looks like an inconsistency/oversight.

image

Note the differences between the description (Lorem ipsum...) position and color on the above text and textarea fields.

### Describe alternatives you've considered

None

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by medavidallsop**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This makes textarea fields in WooCommerce settings position the description (not tooltip based) below the field so consistent with text fields. In addition, it also ensures the color of the description is consistent with the color of descriptions of other fields.

**Note: Since the above was written it has been discussed to instead allow the use of the `desc_at_end` attribute like other fields on textareas, to ensure it doesn't effect any existing textarea fields, and if a dev wants to add the description at the bottom they can by passing the `desc_at_end` attribute as true when registering the textarea setting.**

Closes #51436.

### How to test the changes in this Pull Request:

**Note: These test instructions have been amended as per the changes in the note above.**

1. Checkout this branch
2. Add the following code snippet to `functions.php`:

```
function settings_field_testing( $settings, $current_section ) {

	if ( '' == $current_section ) {

		$settings[] = array(
			'name'	=> 'Some section',
			'id'	=> 'some-section',
			'type'	=> 'title',
		);

		$settings[] = array(
			'name'		=> 'Test text',
			'id'		=> 'test-text',
			'type'		=> 'text',
			'desc'		=> 'Lorem ipsum dolor sit amet.',
		);

		$settings[] = array(
			'name'		=> 'Test text area (description above)',
			'id'		=> 'test-text-area',
			'type'		=> 'textarea',
			'desc'		=> 'Lorem ipsum dolor sit amet.',
			'desc_at_end'	=> false,
		);

		$settings[] = array(
			'name'		=> 'Test text area (description below)',
			'id'		=> 'test-text-area',
			'type'		=> 'textarea',
			'desc'		=> 'Lorem ipsum dolor sit amet.',
			'desc_at_end'	=> true,
		);

		$settings[] = array(
			'type'	=> 'sectionend',
		);

	}

	return $settings;

}
add_filter( 'woocommerce_get_settings_products', 'settings_field_testing', 10, 2 );
```

3. Navigate to WooCommerce > Settings > Products > General, scroll to the bottom, you'll see a test text and textarea field in "Some section"

4. See that the description of the textarea is displayed below the field when `desc_at_end` is `true`

5. See that the description of the textarea is displayed above the field when `desc_at_end` is `false`

6. Remove the `desc_at_end` entirely and see the description is displayed as it is in the current Woo version before these changes

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Message Add conditions so desc_at_end attribute can be used to display field descriptions below textarea fields in settings

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

**Comment by coreymckrill** [comment]
Hey @medavidallsop thanks for the PR! The way this is written right now, it has the potential to break back compatibility because some plugins may be writing code and styles with the assumption that the textarea description is at the beginning. However, I think that if you implemented the additional settings property `desc_at_end` like [the Radio input has](https://github.com/woocommerce/woocommerce/blob/657eaf1bcf1821eaa1af81facdc12ea8c43ca57a/plugins/woocommerce/includes/admin/class-wc-admin-settings.php#L451-L451), you could maintain back compatibility while allowing for a more consistent experience.

**Comment by medavidallsop** [comment]
@coreymckrill Good points, I've gone ahead and amended this to add in conditions to use the `desc_at_end` attribute if you'd like to review.

I have also updated the original test instructions with details of this and amended the steps/code example.

~Also I'm unsure on how you amend the changelog text, I amended the message in the original PR info to "Add conditions so desc_at_end attribute can be used to display field descriptions below textarea fields in settings" but unsure if I need to do anything else, not sure if a commit gets auto generated from this, or I need to manually do something?~ The automation appears to have done this now.

**Comment by medavidallsop** [comment]
@coreymckrill I also just wanted highlight here that the change to this line:

`if ( $description && in_array( $value['type'], array( 'radio' ), true ) ) {` in `get_field_description`

This has had `'textarea'` removed in my commit, this ensures the description is styled consistently, if `textarea` is in this array it makes the description text color different (black rather than the usual gray).

However this does cause the `margin-top:0` to be there in both above/below states. Ideally I'd want to not include that inline CSS if `desc_at_end` is false, as it adds a tiny bit of margin I don't feel is required in that state. However to achieve that it would require passing of `desc_at_end` to `get_field_description`, and unsure of consequences of doing so... just something to consider when you review this.

**Comment by coreymckrill** [comment]
Hey @medavidallsop, this is looking pretty good...

> However this does cause the margin-top:0 to be there in both above/below states. Ideally I'd want to not include that inline CSS if desc_at_end is false, as it adds a tiny bit of margin I don't feel is required in that state.

I think you mean that the margin-top:0 _won't_ be there in both states? That's what I see with your changes.

> However to achieve that it would require passing of desc_at_end to get_field_description, and unsure of consequences of doing so... just something to consider when you review this.

It looks like `desc_at_end` does get passed to that method, as the poorly named `$value` param there is actually the array of all the settings properties...

**Comment by medavidallsop** [comment]
@coreymckrill margin-top won't be there in both states yes.

So that means if desc at end is false, you get a little gap like this (in red):

image

Which ideally shouldn't be there, in that state, but if the description was below, it should be (which it will).

So do you think based on this I should amend `get_field_description` to check for if it is a textarea and conditionally do the margin-top:0 based on the show_desc_at_end?

**Comment by coreymckrill** [comment]
> So do you think based on this I should amend get_field_description to check for if it is a textarea and conditionally do the margin-top:0 based on the show_desc_at_end?

Yeah, I think it would be reasonable to add a separate if/then case in `get_field_description` for textarea so it can by optimally styled.

**Comment by medavidallsop** [comment]
@coreymckrill I have just submitted a change related to that, if you can check it over. Thank you!

File: plugins/woocommerce/includes/admin/class-wc-admin-settings.php [hunk 1]
```diff
@@ -380,15 +380,19 @@ class="<?php echo esc_attr( $value['class'] ); ?>colorpick"
 					// Textarea.
 					case 'textarea':
 						$option_value = $value['value'];
+						$show_desc_at_end = $value['desc_at_end'] ?? false;
```
Comment (coreymckrill) [inline plugins/woocommerce/includes/admin/class-wc-admin-settings.php]:
```suggestion
						$option_value     = $value['value'];
						$show_desc_at_end = $value['desc_at_end'] ?? false;
```

Make the code linter happy

**Review by coreymckrill — APPROVED** [review]
👍 Works well, and seems like it should be completely backward compatible. Thanks for the PR @medavidallsop !

**Pull request merged**